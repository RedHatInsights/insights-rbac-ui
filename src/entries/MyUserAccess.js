import React, { lazy, Suspense, useEffect } from 'react';
import { BrowserRouter as Router, Route, useHistory } from 'react-router-dom';
import { Provider } from 'react-redux';
import { getBaseName } from '@redhat-cloud-services/frontend-components-utilities/helpers';
import Main from '@redhat-cloud-services/frontend-components/Main';
import { IntlProvider } from 'react-intl';

import registry, { RegistryContext } from '../utilities/store';
import messages from '../locales/data.json';
import ErroReducerCatcher from '../presentational-components/shared/ErrorReducerCatcher';
import PermissionsContext from '../utilities/permissions-context';
import pathnames from '../utilities/pathnames';
import { AppPlaceholder } from '../presentational-components/shared/loader-placeholders';
import useUserData from '../hooks/useUserData';

const MyUserAccess = lazy(() => import('../smart-components/myUserAccess/MUAHome'));

import '../App.scss';

export const locale = 'en';

const Routes = () => {
  const history = useHistory();
  useEffect(() => {
    // redirect to MUA if url is "/settings"
    if (window?.location?.pathname?.match(/\/settings$/)) {
      history.push('/my-user-access');
    }
    // set correct title
    document.title = 'My User Access';
  }, []);

  return (
    <Suspense fallback={<AppPlaceholder />}>
      <Route path={pathnames['my-user-access'].path} component={MyUserAccess} />
    </Suspense>
  );
};

const MuaApp = () => {
  const userData = useUserData();
  return (
    <PermissionsContext.Provider value={{ ...userData }}>
      <ErroReducerCatcher>
        <Main style={{ marginLeft: 0, padding: 0 }}>
          <Routes />
        </Main>
      </ErroReducerCatcher>
    </PermissionsContext.Provider>
  );
};

const SettingsMua = () => (
  <IntlProvider locale={locale} messages={messages[locale]}>
    <RegistryContext.Provider
      value={{
        getRegistry: () => registry,
      }}
    >
      <Provider store={registry.getStore()}>
        <Router basename={getBaseName(location.pathname, 1)}>
          <MuaApp />
        </Router>
      </Provider>
    </RegistryContext.Provider>
  </IntlProvider>
);

export default SettingsMua;
