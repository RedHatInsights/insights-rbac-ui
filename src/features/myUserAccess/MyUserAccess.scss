.rbac-p-myUserAccess {
  &--title {
    display: flex;      
    align-items: center;
    background-color: var(--pf-v5-global--BackgroundColor--100);
    padding: var(--pf-v5-global--spacer--lg) 0 var(--pf-v5-global--spacer--lg) var(--pf-v5-global--spacer--lg);
    .pf-v5-c-label { margin-left: var(--pf-v5-global--spacer--md); }
  }
  &--subtitle {
    position: relative;
    background-color: var(--pf-v5-global--BackgroundColor--100);
    box-shadow: var(--pf-v5-global--BoxShadow--md-bottom);
    padding: 0  var(--pf-v5-global--spacer--lg) var(--pf-v5-global--spacer--md) var(--pf-v5-global--spacer--lg);
  }
  &--dropdown {
    display: none;
  }
}

.rbac-l-myUserAccess-section__cards,
.rbac-l-myUserAccess-section__table {
  background-color: var(--pf-v5-global--BackgroundColor--100);
  height: calc(100vh - 111px - 84px);
  padding: 0 var(--pf-v5-global--spacer--lg);
  overflow-x: auto;
  h3 {padding: var(--pf-v5-global--spacer--lg) 0;}
}
.rbac-l-myUserAccess-section__cards {border-right: 1px solid var(--pf-v5-global--BorderColor--100);}

@media only screen and (max-width: 1200px) {
  .sticky {
    position: sticky;
    position: -webkit-sticky;
    background-color: var(--pf-v5-global--BackgroundColor--100);
    height: 60px;
    width: 100%;
    z-index: 200;
  }

  .rbac-p-myUserAccess {
    &--title {
      padding: var(--pf-v5-global--spacer--xl) var(--pf-v5-global--spacer--lg) var(--pf-v5-global--spacer--md) var(--pf-v5-global--spacer--lg);
      top: 0;
    }
    &--subtitle {
      box-shadow: none;
      padding: 0 var(--pf-v5-global--spacer--lg) var(--pf-v5-global--spacer--lg) var(--pf-v5-global--spacer--lg);
    } 
    &--dropdown {
      display: block;
      padding: 0 var(--pf-v5-global--spacer--lg);
      top: 60px;
      .pf-v5-c-dropdown {
        width: 100%;
        button {
          width: 100%;
        }
      }
    }
    &-section__table {
      height: auto;
      margin: var(--pf-v5-global--spacer--md);
      > .pf-v5-c-title { display: none; }
    }
  }

  .rbac-l-myUserAccess-section__table {
    background-color: transparent;
    h3 {
      background-color: var(--pf-v5-global--BackgroundColor--100);
      margin-top: var(--pf-v5-global--spacer--md);
      padding-left: var(--pf-v5-global--spacer--md);
    }
  }
}
