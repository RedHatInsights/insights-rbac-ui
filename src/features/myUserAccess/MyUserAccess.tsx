import React, { useContext, useEffect, useState } from 'react';
import { NavLink, useLocation, useSearchParams } from 'react-router-dom';
import { Dropdown, DropdownItem, DropdownToggle } from '@patternfly/react-core/deprecated';

import { PageHeaderTitle } from '@redhat-cloud-services/frontend-components/PageHeader';
import { useChrome } from '@redhat-cloud-services/frontend-components/useChrome';
import { Spinner, Text } from '@patternfly/react-core';
import StatusLabel from './components/StatusLabel';
import PermissionsContext from '../../utilities/permissionsContext';
import { bundleData } from './bundleData';
import { UserAccessLayout } from './components/UserAccessLayout';
import type { Entitlements } from './components/BundleCard';
import { AccessTable } from './AccessTable';
import { RolesTable } from './RolesTable';
import OrgAdminContext from '../../utilities/orgAdminContext';
import { useBundleApps } from './useBundleApps';
import { DEFAULT_MUA_BUNDLE } from '../../utilities/constants';
import { useIntl } from 'react-intl';
import messages from '../../Messages';

import './MyUserAccess.scss';

interface UserEntitlements {
  [key: string]: { is_entitled: boolean };
}

interface User {
  entitlements?: UserEntitlements;
  isOrgAdmin?: boolean;
}

// FilterState is now in shared types.ts file

export const MyUserAccess: React.FC = () => {
  const intl = useIntl();
  const chrome = useChrome();
  const location = useLocation();
  const [user, setUser] = useState<User>({});
  const [searchParams, setSearchParams] = useSearchParams();
  const bundle = searchParams.get('bundle');
  const { userAccessAdministrator } = useContext(PermissionsContext);

  useEffect(() => {
    chrome.auth.getUser().then((user: any) => {
      const { identity, entitlements } = user || {};
      setUser({
        entitlements,
        isOrgAdmin: identity?.user?.is_org_admin,
      });
    });

    // Only set default bundle if bundle is falsy and we haven't already set it
    if (!bundle) {
      setSearchParams({ bundle: DEFAULT_MUA_BUNDLE });
    }
  }, [bundle, setSearchParams]);

  const enhancedEntitlements: UserEntitlements = {
    ...user.entitlements,
  };

  const entitledBundles: Entitlements = Object.entries(enhancedEntitlements).filter(([, { is_entitled }]) => is_entitled);

  const [isDropdownOpen, setDropdownOpen] = useState(false);

  // Bundle and filter logic moved from BundleView
  const apps = useBundleApps(bundle || undefined);
  const hasAdminAccess = user.isOrgAdmin || userAccessAdministrator;

  // No more bloated FilterItem[] at business logic level!

  return (
    <React.Fragment>
      {Object.prototype.hasOwnProperty.call(user, 'entitlements') &&
      (Object.prototype.hasOwnProperty.call(user, 'isOrgAdmin') || userAccessAdministrator) ? (
        <React.Fragment>
          <PageHeaderTitle
            className="rbac-p-myUserAccess--title sticky"
            title={
              <React.Fragment>
                <span> {intl.formatMessage(messages.myUserAccess)} </span>
                <StatusLabel isOrgAdmin={user.isOrgAdmin} isUserAccessAdmin={userAccessAdministrator} />
              </React.Fragment>
            }
          />
          <Text component="p" className="rbac-p-myUserAccess--subtitle">
            {intl.formatMessage(messages.selectAppsToViewPermissions)}
          </Text>
          {entitledBundles && (
            <div className="rbac-p-myUserAccess--dropdown sticky">
              <Dropdown
                ouiaId="mua-bundle-dropdown"
                toggle={
                  <DropdownToggle onToggle={() => setDropdownOpen(!isDropdownOpen)} id="mua-bundle-dropdown">
                    {bundle
                      ? bundleData.find(({ entitlement }) => entitlement === bundle)?.title
                      : intl.formatMessage(messages.chooseSubscriptionEllipsis)}
                  </DropdownToggle>
                }
                dropdownItems={bundleData.map((data) => (
                  <NavLink key={data.entitlement} to={{ pathname: location.pathname, search: `bundle=${data.entitlement}` }}>
                    <DropdownItem
                      onClick={() => {
                        setDropdownOpen(false);
                      }}
                      component="button"
                    >
                      {data.title}
                    </DropdownItem>
                  </NavLink>
                ))}
                isOpen={isDropdownOpen}
              />
            </div>
          )}
          <section>
            <UserAccessLayout
              entitledBundles={entitledBundles}
              title={intl.formatMessage(user.isOrgAdmin || userAccessAdministrator ? messages.yourRoles : messages.yourPermissions, {
                name: bundleData.find(({ entitlement }) => entitlement === (bundle || DEFAULT_MUA_BUNDLE))?.title,
              })}
              currentBundle={bundle || DEFAULT_MUA_BUNDLE}
            >
              <OrgAdminContext.Provider value={hasAdminAccess}>
                {hasAdminAccess ? (
                  <RolesTable key={bundle} apps={apps} showResourceDefinitions={hasAdminAccess} />
                ) : (
                  <AccessTable key={bundle} apps={apps} showResourceDefinitions={hasAdminAccess} />
                )}
              </OrgAdminContext.Provider>
            </UserAccessLayout>
          </section>
        </React.Fragment>
      ) : (
        <Spinner />
      )}
    </React.Fragment>
  );
};

export default MyUserAccess;
