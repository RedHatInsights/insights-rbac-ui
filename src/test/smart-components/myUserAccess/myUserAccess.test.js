import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import MyUserAccess from '../../../smart-components/myUserAccess/myUserAccess';

describe('<MyUserAccess />', () => {
  it('should render correctly', () => {
    expect(toJson(mount(
        <MyUserAccess/>
    ))).toMatchSnapshot();
  });
});
