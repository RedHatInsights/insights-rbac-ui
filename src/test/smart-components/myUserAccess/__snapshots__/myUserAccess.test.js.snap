// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MyUserAccess /> should render correctly 1`] = `
<MyUserAccess>
  <Spinner>
    <span
      aria-valuetext="Loading..."
      className="pf-c-spinner pf-m-xl"
      role="progressbar"
    >
      <span
        className="pf-c-spinner__clipper"
      />
      <span
        className="pf-c-spinner__lead-ball"
      />
      <span
        className="pf-c-spinner__tail-ball"
      />
    </span>
  </Spinner>
</MyUserAccess>
`;
